Fix torch.cuda.check_error type errors

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41330 Fix torch.cuda.check_error type errors**

`torch.cuda.check_error` is annotated as taking an `int` as argument but when running `torch.cuda.check_error(34)` one would get:
```
TypeError: cudaGetErrorString(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch._C._cudart.cudaError) -> str

Invoked with: 34
```
Even if one explicitly casted the argument, running `torch.cuda.check_error(torch._C._cudart.cudaError(34))` would give:
```
AttributeError: 'str' object has no attribute 'decode'
```

This PR fixes both issues (thus allowing `check_error` to be called with a un-casted int) and adds a test.

Differential Revision: [D22500549](https://our.internmc.facebook.com/intern/diff/D22500549/)

lw committed:
Fix torch.cuda.check_error type errors

`torch.cuda.check_error` is annotated as taking an `int` as argument but when running `torch.cuda.check_error(34)` one would get:
```
TypeError: cudaGetErrorString(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch._C._cudart.cudaError) -> str

Invoked with: 34
```
Even if one explicitly casted the argument, running `torch.cuda.check_error(torch._C._cudart.cudaError(34))` would give:
```
AttributeError: 'str' object has no attribute 'decode'
```

This PR fixes both issues (thus allowing `check_error` to be called with a un-casted int) and adds a test.

Differential Revision: [D22500549](https://our.internmc.facebook.com/intern/diff/D22500549/)

[ghstack-poisoned]

lw commented:
This code seems to last have been changed in #33678, so I'm adding the author and reviewer from that PR as reviewers here: @peterbell10, @ezyang.

lw commented:
Also, should this fix be cherry-picked to v1.6?

lw committed:
Update on "Fix torch.cuda.check_error type errors"


`torch.cuda.check_error` is annotated as taking an `int` as argument but when running `torch.cuda.check_error(34)` one would get:
```
TypeError: cudaGetErrorString(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch._C._cudart.cudaError) -> str

Invoked with: 34
```
Even if one explicitly casted the argument, running `torch.cuda.check_error(torch._C._cudart.cudaError(34))` would give:
```
AttributeError: 'str' object has no attribute 'decode'
```

This PR fixes both issues (thus allowing `check_error` to be called with a un-casted int) and adds a test.

Differential Revision: [D22500549](https://our.internmc.facebook.com/intern/diff/D22500549/)

[ghstack-poisoned]

peterbell10 reviewed (approved):
:+1:

lw commented:
The failure on ROCm seems relevant:
```
10:59:47 ======================================================================
10:59:47 ERROR: test_check_error (__main__.TestCuda)
10:59:47 ----------------------------------------------------------------------
10:59:47 Traceback (most recent call last):
10:59:47   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 777, in wrapper
10:59:47     method(*args, **kwargs)
10:59:47   File "test_cuda.py", line 311, in test_check_error
10:59:47     torch.cuda.check_error(2)
10:59:47   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/cuda/__init__.py", line 225, in check_error
10:59:47     raise CudaError(res)
10:59:47   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/cuda/__init__.py", line 219, in __init__
10:59:47     msg = _cudart.cudaGetErrorString(torch._C._cudart.cudaError(code))
10:59:47 AttributeError: module 'torch._C._cudart' has no attribute 'cudaGetErrorString'
```
I'm suspecting this is because of the HIP rewrite rules (which I only just found out about), which mean that therefore we should actually call `hipGetErrorString`.

However, I don't think the Python code should have to deal with: it should probably just call one function with a consistent name that gets resolved by the C++ to the two different versions based on the compile-time flags. The problem is that the rewrite is naive and also changes the name of the Python function. We should probably use a different name, perhaps just without the leading `cuda` prefix.

lw committed:
Update on "Fix torch.cuda.check_error type errors"


`torch.cuda.check_error` is annotated as taking an `int` as argument but when running `torch.cuda.check_error(34)` one would get:
```
TypeError: cudaGetErrorString(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch._C._cudart.cudaError) -> str

Invoked with: 34
```
Even if one explicitly casted the argument, running `torch.cuda.check_error(torch._C._cudart.cudaError(34))` would give:
```
AttributeError: 'str' object has no attribute 'decode'
```

This PR fixes both issues (thus allowing `check_error` to be called with a un-casted int) and adds a test.

Differential Revision: [D22500549](https://our.internmc.facebook.com/intern/diff/D22500549/)

[ghstack-poisoned]

lw committed:
Update on "Fix torch.cuda.check_error type errors"


`torch.cuda.check_error` is annotated as taking an `int` as argument but when running `torch.cuda.check_error(34)` one would get:
```
TypeError: cudaGetErrorString(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch._C._cudart.cudaError) -> str

Invoked with: 34
```
Even if one explicitly casted the argument, running `torch.cuda.check_error(torch._C._cudart.cudaError(34))` would give:
```
AttributeError: 'str' object has no attribute 'decode'
```

This PR fixes both issues (thus allowing `check_error` to be called with a un-casted int) and adds a test.

Differential Revision: [D22500549](https://our.internmc.facebook.com/intern/diff/D22500549/)

[ghstack-poisoned]

peterbell10 commented:
Shouldn't a fairly simple workaround be to split the function name into two strings, e.g.
```c++
  cudart.def("cuda" "GetErrorString", cudaGetErrorString);
```

That would preserve the names and workaround the cuda to hip renaming.

lw commented:
Nice trick, I'll do that.

lw committed:
Update on "Fix torch.cuda.check_error type errors"


`torch.cuda.check_error` is annotated as taking an `int` as argument but when running `torch.cuda.check_error(34)` one would get:
```
TypeError: cudaGetErrorString(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch._C._cudart.cudaError) -> str

Invoked with: 34
```
Even if one explicitly casted the argument, running `torch.cuda.check_error(torch._C._cudart.cudaError(34))` would give:
```
AttributeError: 'str' object has no attribute 'decode'
```

This PR fixes both issues (thus allowing `check_error` to be called with a un-casted int) and adds a test.

Differential Revision: [D22500549](https://our.internmc.facebook.com/intern/diff/D22500549/)

[ghstack-poisoned]

lw committed:
Update on "Fix torch.cuda.check_error type errors"


`torch.cuda.check_error` is annotated as taking an `int` as argument but when running `torch.cuda.check_error(34)` one would get:
```
TypeError: cudaGetErrorString(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch._C._cudart.cudaError) -> str

Invoked with: 34
```
Even if one explicitly casted the argument, running `torch.cuda.check_error(torch._C._cudart.cudaError(34))` would give:
```
AttributeError: 'str' object has no attribute 'decode'
```

This PR fixes both issues (thus allowing `check_error` to be called with a un-casted int) and adds a test.

Differential Revision: [D22500549](https://our.internmc.facebook.com/intern/diff/D22500549/)

[ghstack-poisoned]

ezyang reviewed (approved):
The HIP workaround is kind of goofy but fine.

lw commented:
@ezyang Should I cherry-pick this on v1.6? (or, if too late, for v1.6.1?)
